### DISCUSSION TITLE
CRUD relational data

### DISCUSSION THREAD

**Discussion by gazialankus** [serverpod/serverpod#3008] [Q&A]
What is the proper way to delete relational data? When I try to delete the main row I get an exception because there is a foreign key relation. So I'm assuming I need to delete that row that the foreign key points to in the other table, and then delete this one here? 

Or do I need to detach in the main table, delete in the foreign table, and then delete in the main table? 

Docs could use some help in this regard. Some clarification about deletions and relations would be nice to have in documents here: https://docs.serverpod.dev/concepts/database/relation-queries#update

Maybe could also be mentioned in the CRUD page here? https://docs.serverpod.dev/concepts/database/crud#delete

I'm not sure what exactly cascade does here. https://docs.serverpod.dev/concepts/database/relations/referential-actions 
I deleted from main but the data in foreign table did not get deleted. Maybe an example there would help.

**Comment by gazialankus** [comment]
I think the relations section needs CRUD examples for each kind of relation. https://docs.serverpod.dev/concepts/database/relations/one-to-one It would be awesome to have a kitchen sink example with all kinds of relations experienced in working code in various cases. Please let me know if such an example actually exists. I think such a comprehensive example would help with Serverpod adoption.

Something that I ran into today that makes my code feel ugly: I have main table with secondaryId pointing to a secondary table. When the secondaryId is non-nullable, 
- In the client, you can't put together a new main-secondary object pair and send it back to server to be added to the database. I'm saving -1 for secondaryId along with the new secondary object, so I can actually send the two to the server. Feels hacky.
- In the server, you can't save the main record before fully storing secondary, since main needs a non-nullable secondaryId. So you have to store the secondary record, get the id of this new record and put it into the main record's non-nullable secondaryId. Feels inefficient. 

I think UUID-like keys would make such situations a bit better by creating the unique key without going to the database. It probably is a big ask at the moment, but I think these situations that people could find "ugly" should be made a bit clearer in docs.

**Comment by gazialankus** [comment]
To solidify my original question about delete: 

```
class: User
table: user
fields:
  userInfo: module:auth:UserInfo?, relation
indexes:
  user_info_id_unique_idx:
    fields: userInfoId
    unique: true
```

This is a relation with the UserInfo table from the auth module.

I can delete records of my User, but the corresponding record in UserInfo stays.

I cannot delete the record from UserInfo before deleting the corresponding User record, I get this exception.
```
DatabaseException: update or delete on table "serverpod_user_info" violates foreign key constraint "user_fk_0" on table "user"
```

So I need to delete User, followed by UserInfo. I need to issue two delete queries.

Then, what does the default Cascade do? https://docs.serverpod.dev/concepts/database/relations/referential-actions

**Comment by SandPod** [comment]
Hey @gazialankus! 👋 

### Relational delete

> What is the proper way to delete relational data

The referential actions are intended to do precisely what you are requesting. Removing the "parent" record should also remove any related tables.

But I think there are two issues here.

#### "Parent" record

I think there is some confusion around what is considered the "parent" record.

In your case, you have created a relation to a module. Since the framework currently does not allow modifications to modules from a project, the module model will be considered the parent, and your project model will be considered the child.

If the referential action is set to `onDelete=cascade`, removing the `UserInfo` entry would remove the `User` entry.

#### Default referential actions

When inspecting our code it seems like there are two different sets of default values for the `OnDelete` refential action.

If the field is defined as an id relation, the default value is `Cascade`.
```dart
class: Example
table: example
fields:
  parentId: int?, relation(parent=example)
```

If the field is defined as an object relation, the default is `NoAction`.
```dart
class: Example
table: example
fields:
  parent: Example?, relation
```

I'll create an issue to document this correctly.

### Additional note

> It would be awesome to have a kitchen sink example with all kinds of relations experienced in working code in various cases.

I agree, I think that would be really helpful. Do you mind creating a "documentation" issue on the Serverpod Repo and list down some of the cases that you would like to see already? Then we can populate the list with more examples and make a work item out of it.

**Comment by SandPod** [comment]
PR for documenting the default values for referential actions here: https://github.com/serverpod/serverpod_docs/pull/221
